src/resmom/mom_walltime.c

```diff
@@ -136,4 +136,47 @@ stop_walltime(job *pjob) {
     /* update walltime and stop accumulating */
     update_walltime(pjob);
     pjob->ji_walltime_stamp = 0;
+}
+
+/**
+ * @brief
+ *
+ *		This function tries to recover the used walltime of a job.
+ *
+ * @param[in] 	pjob	    - pointer to the job
+ *
+ * @return	void
+ *
+ * @par MT-safe: No
+ */
+void
+recover_walltime(job *pjob) {
+    attribute       *resources_used;
+    resource_def    *walltime_def;
+    resource        *used_walltime;
+
+    if (NULL == pjob)
+        return;
+
+    if (0 == pjob->ji_qs.ji_stime)
+        return;
+
+    if (0 == time_now)
+        time_now = time(NULL);
+
+    resources_used = &pjob->ji_wattr[(int)JOB_ATR_resc_used];
+    assert(resources_used != NULL);
+    walltime_def = find_resc_def(svr_resc_def, "walltime", svr_resc_size);
+    assert(walltime_def != NULL);
+    used_walltime = find_resc_entry(resources_used, walltime_def);
+
+    /*
+     * if the used walltime is not set, try to recover it.
+     */
+    if (NULL == used_walltime) {
+        used_walltime = add_resource_entry(resources_used, walltime_def);
+        used_walltime->rs_value.at_flags |= ATR_VFLAG_SET;
+        used_walltime->rs_value.at_type = ATR_TYPE_LONG;
+        used_walltime->rs_value.at_val.at_long = (long)((double)(time_now - pjob->ji_qs.ji_stime) * wallfactor);
```

Comment:
`ji_qs.ji_stime` records the start time of the job. `time_now - pjob->ji_qs.ji_stime` does not always yield the job's walltime.

Reply:
The job could have been suspended and `time_now - pjob->ji_qs.ji_stime` will be larger the the actuall walltime.

Reply:
I understand. As mentioned in [PP-1117](https://pbspro.atlassian.net/browse/PP-1117), the accuracy obtained in [PP-772](https://pbspro.atlassian.net/browse/PP-772) is partially lost for recovered jobs. I mentioned an idea to do a regular full save job but this idea was rejected due to performance. Do you have a better idea how to recover the used walltime?

Maybe, we could do the full save job of running jobs only on mom exit? What is your thought?

Reply:
Maybe we could save resources_used.walltime only?